scan_for_updates.py

```diff
@@ -96,76 +113,67 @@ def gap_exec(commands, gap="gap"):
+                warning(
+                    "error trying to download {}, status code {}, skipping!".format(
+                        url, response.status_code
+                    )
+                )
+                return
+            hash_url = hashlib.sha256(response.text.encode("utf-8"))
+            if hash_url != hash_distro:
+                notice(pkg_name + ": detected different sha256 hash")
+                fmt = pkg_json["ArchiveFormats"].split(" ")[0]
+                url = pkg_json["ArchiveURL"] + fmt
+                with open(
+                    os.path.join(pkginfos_dir, pkg_name + ".g"), "w"
+                ) as f:
+                    f.write(response.text)
+                download_archive(pkg_name, url, archive_dir, pkg_name + fmt)
+    except (OSError, IOError):
         notice(pkg_name + ": missing meta.json file, skipping!")
         return
 
-    with open(fname, "r") as f:
-        pkg_json = json.load(f)
-        try:
-            hash_distro = pkg_json["PackageInfoSHA256"]
-        except KeyError:
-            notice(pkg_name + ': missing key "PackageInfoSHA256"')
-            hash_distro = 0
-        url = pkg_json["PackageInfoURL"]
-        response = requests.get(url)
-        if response.status_code != 200:
-            warning(
-                "error trying to download {}, status code {}, skipping!".format(
-                    url, response.status_code
-                )
-            )
-            return
-        hash_url = hashlib.sha256(response.text.encode("utf-8"))
-        if hash_url != hash_distro:
-            notice(pkg_name + ": detected different sha256 hash")
-            fmt = pkg_json["ArchiveFormats"].split(" ")[0]
-            url = pkg_json["ArchiveURL"] + fmt
-            with open(os.path.join(pkginfos_dir, pkg_name + ".g"), "w") as f:
-                f.write(response.text)
-            download_archive(pkg_name, url, archive_dir, pkg_name + fmt)
-
 
-def main():
-    archive_dir = "_archives"
-    json_dir = "_json"
-    pkginfos_dir = "_pkginfos"
+def scan_for_updates(archive_dir, pkginfos_dir):
+    if not os.path.exists(archive_dir):
+        os.mkdir(archive_dir)
+    assert os.path.isdir(archive_dir)
+    if not os.path.exists(pkginfos_dir):
+        os.mkdir(pkginfos_dir)
+    assert os.path.isdir(pkginfos_dir)
     for pkgname in sorted(os.listdir(os.getcwd())):
-        if (
-            not pkgname.startswith(".")
-            and not pkgname.startswith("_")
-            and os.path.isdir(pkgname)
-        ):
-            scan_for_updates(archive_dir, pkginfos_dir, pkgname)
+        if not skip_dir(pkgname):
+            scan_for_one_update(archive_dir, pkginfos_dir, pkgname)
+
 
+def output_json(pkginfos_dir):
     if (
         gap_exec(
-            r"OutputJson(\"{}\", \"{}\");".format(pkginfos_dir, json_dir),
-            gap="gap scan_for_updates.g",
+            r"OutputJson(\"{}\");".format(pkginfos_dir),
+            gap="gap _tools/scan_for_updates.g",
```

Comment:
There is no such directory? Or maybe you assume it is there and contains a clone of this repo?

Reply:
Fixed